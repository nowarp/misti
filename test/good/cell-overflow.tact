fun refs_num_bad(c: Cell) {
    let b1 = beginCell(); // Bad: 4+ refs
    b1 = b1.storeRef(c);
    b1 = b1.storeRef(c);
    b1 = b1.storeRef(c);
    b1 = b1.storeRef(c);
    b1 = b1.storeRef(c);
    let c1 = b1.endCell();

    let c2 = // Bad: 4+ refs
    beginCell()
        .storeRef(c)
        .storeRef(c)
        .storeRef(c)
        .storeRef(c)
        .storeRef(c)
        .endCell();
}

fun store_size_bad() {
    let b1 = beginCell(); // Bad: 1023+ bits
    b1 = b1.storeInt(42, 257);
    b1 = b1.storeInt(42, 259-2);
    b1 = b1.storeInt(42, 256+1);
    b1 = b1.storeInt(42, 257);
    let c1 = b1.endCell();

    let b2 = beginCell(); // Bad: 1023+ bits
    b2 = b2.storeInt(1, 257)
           .storeInt(1, 257)
           .storeInt(1, 257)
           .storeInt(1, 210); // Total size: 981
    b2 = b2.storeCoins(40000000000); // 44 + 981 = 1025 => overflow
    c1 = b2.endCell();

    let b_1000_bits = beginCell()
           .storeInt(1, 257)
           .storeInt(1, 257)
           .storeInt(1, 257)
           .storeInt(1, 229);
    let b3 = beginCell(); // Bad: 1025 bits
    b3 = b3.storeBuilder(b_1000_bits)
           .storeInt(1, 25);
    c1 = b3.endCell();
}

fun refs_num_ok(c: Cell, copt: Cell?) {
    let c1 = // OK: Cell could be None (no taint analysis yet)
    beginCell()
        .storeRef(c)
        .storeRef(c)
        .storeRef(c)
        .storeMaybeRef(copt)
        .storeRef(c)
        .endCell();

}

