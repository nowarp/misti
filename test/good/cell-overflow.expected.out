[CRITICAL] CellOverflow: Maximum number of references exceeded
test/good/cell-overflow.tact:2:9:
  1 | fun refs_num_bad(c: Cell) {
> 2 |     let b1 = beginCell(); // Bad: 4+ refs
              ^
  3 |     b1 = b1.storeRef(c);
Cells cannot contain more than 4 references to other cells
Help: Remove extra .storeRef operations
See: https://nowarp.io/tools/misti/docs/detectors/CellOverflow

[CRITICAL] CellOverflow: Maximum amount of data stored in cell exceeded
test/good/cell-overflow.tact:21:9:
  20 | fun store_size_bad() {
> 21 |     let b1 = beginCell(); // Bad: 1023+ bits
               ^
  22 |     b1 = b1.storeInt(42, 257);
Cells cannot store more than 1023 bits (expected size: 1028)
Help: Remove extra .store operations
See: https://nowarp.io/tools/misti/docs/detectors/CellOverflow

[CRITICAL] CellOverflow: Maximum amount of data stored in cell exceeded
test/good/cell-overflow.tact:28:9:
  27 | 
> 28 |     let b2 = beginCell(); // Bad: 1023+ bits
               ^
  29 |     b2 = b2.storeInt(1, 257)
Cells cannot store more than 1023 bits (expected size: 1025)
Help: Remove extra .store operations
See: https://nowarp.io/tools/misti/docs/detectors/CellOverflow

[CRITICAL] CellOverflow: Maximum amount of data stored in cell exceeded
test/good/cell-overflow.tact:41:9:
  40 |            .storeInt(1, 229);
> 41 |     let b3 = beginCell(); // Bad: 1025 bits
               ^
  42 |     b3 = b3.storeBuilder(b_1000_bits)
Cells cannot store more than 1023 bits (expected size: 1025)
Help: Remove extra .store operations
See: https://nowarp.io/tools/misti/docs/detectors/CellOverflow

[MEDIUM] NeverAccessedVariables: Variable c1 is never accessed
test/good/cell-overflow.tact:8:5:
  7 |     b1 = b1.storeRef(c);
> 8 |     let c1 = b1.endCell();
          ^
  9 | 
Help: Consider removing the variable
See: https://nowarp.io/tools/misti/docs/detectors/NeverAccessedVariables

[MEDIUM] NeverAccessedVariables: Variable c2 is never accessed
test/good/cell-overflow.tact:10:5:
   9 | 
> 10 |     let c2 = // Bad: 4+ refs
           ^
  11 |     beginCell()
Help: Consider removing the variable
See: https://nowarp.io/tools/misti/docs/detectors/NeverAccessedVariables

[MEDIUM] NeverAccessedVariables: Write-only variable: c1
test/good/cell-overflow.tact:26:5:
  25 |     b1 = b1.storeInt(42, 257);
> 26 |     let c1 = b1.endCell();
           ^
  27 | 
Help: The variable value should be accessed
See: https://nowarp.io/tools/misti/docs/detectors/NeverAccessedVariables

[MEDIUM] NeverAccessedVariables: Variable c1 is never accessed
test/good/cell-overflow.tact:48:5:
  47 | fun refs_num_ok(c: Cell, copt: Cell?) {
> 48 |     let c1 = // OK: Cell could be None (no taint analysis yet)
           ^
  49 |     beginCell()
Help: Consider removing the variable
See: https://nowarp.io/tools/misti/docs/detectors/NeverAccessedVariables